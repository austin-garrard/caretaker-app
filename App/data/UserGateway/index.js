export class Permissions {
    static FOCUS = 'focus';
    static ADMIN = 'admin';
    static HELPER = 'helper';
}

export default class UserGateway {

    static currentUser = null;
    static allUsers = null;

    static getAll() {
        if(this.allUsers === null) {
            //api call goes here
            this.allUsers = [
                {name: 'Sarah', identifier: 'sarah@emailprovider.com', permission: Permissions.FOCUS, roles: [], phone:'123-456-7890'},
                {name: 'Sarah', identifier: 's', permission: Permissions.FOCUS, roles: [], phone:'123-456-7890'},
                {name: 'Caroline', identifier: 'caroline@woahdude.com', permission: Permissions.ADMIN, roles: ['driver', 'coordinator'], phone:'123-456-7890'},
                {name: 'Jack', identifier: 'jack@coolwebsite.com', permission: Permissions.HELPER, roles: ['active friend']},
                {name: 'Austin', identifier: 'austin@yeehaw.com', permission: Permissions.HELPER, roles: [], phone:'123-456-7890'}
            ]
        }
        return this.allUsers;
    }

    static getName(identifier) {
        const user = this.allUsers.find((user) => user.identifier === identifier)
        return user ? user.name : 'TBD';
    }

    static getRoles() {
        return this.currentUser.roles
    }

    static signIn(email) {
        //make call to google oauth and store the results
        let results = {
            email: email
        };
        let userFromBackend = this.getAll().find((user) => user.identifier === results.email);
        if(typeof userFromBackend !== 'undefined') {
            this.currentUser = userFromBackend;
            return true;
        }
        return false;
    }

    static isSelf(identifier) {
        return this.currentUser.identifier === identifier;
    }

    static isAdmin() {
        return this.currentUser.permission === Permissions.ADMIN
            || this.currentUser.permission === Permissions.FOCUS;
    }

    static getCurrentPermissions() {
        return this.currentUser.permission;
    }

    static inviteUser(email) {
        //api call to invite the user
        alert('invited ' + email + '!');
    }
}
