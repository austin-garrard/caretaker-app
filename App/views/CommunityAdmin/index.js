import React, {Component, PureComponent} from 'react';
import {
  StyleSheet,
  Text,
  View,
  FlatList,
  Button
} from 'react-native';
import ContainerFor from './container.js';

const ListItem = (props) => (
    <View style={styles.person}>
    <View style={styles.item}>
        <Text>{props.name}</Text>
        <Text>{props.email}</Text>
        <Text>{props.phone}</Text>
    </View>
    <View>
        <Text style={styles.roles}>{props.roles.join(', ')}</Text>
    </View>
    </View>
)

const CommunityAdmin = (props) => {
    _renderItem = ({item}) => (
        <ListItem
            name={item.name}
            email={item.identifier}
            phone={item.phone}
            roles={item.roles}
        />
    )

    return (
        <View style={styles.container}>
            {props.modal}
            <FlatList
               data={props.users}
               renderItem={_renderItem}
               keyExtractor={(item) => item.identifier}
            />
            <Button
                style={styles.button}
                title='Invite'
                onPress={props.onShowModal}
            />
        </View>
    );
}

export default ContainerFor(CommunityAdmin);

const styles = StyleSheet.create({
  button: {
    margin: 10,
    padding: 10
  },
  container: {
    justifyContent: 'flex-start',
    backgroundColor: '#F5FCFF',
  },
  title: {
    fontSize: 24,
    textAlign: 'center',
    margin: 15,
  },
  person: {
    padding: 8,
  },
  item: {
    flexDirection: 'row',
    flex: 1,
    justifyContent: 'space-between',
  },
  roles: {
    flexDirection: 'row',
    textAlign: 'center',
    flex: 1,
    justifyContent: 'space-between',
  }
});
