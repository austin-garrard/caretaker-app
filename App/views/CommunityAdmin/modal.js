import React from 'react';
import {View, TextInput, Button, Modal} from 'react-native';

export default CommunityAdminModal = (props) => {

    return (
        <Modal visible={props.visible}
            animationType={"fade"}
            transparent={false}
            onRequestClose={() => {alert("Modal has been closed.")}}>
            <View>
                <TextInput
                    onChangeText={props.onUpdateEmail}
                    placeholder='Enter email'/>
                <Button title="Submit Email" onPress={props.onSubmitEmail} />
                <Button title="Cancel" onPress={props.onHideModal} />
            </View>
        </Modal>
    );
}
