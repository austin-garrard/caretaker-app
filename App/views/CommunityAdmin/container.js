import React, {Component} from 'react';
import {View} from 'react-native';
import UserGateway from '../../data/UserGateway/';
import CommunityAdminModal from './modal.js';

export default ContainerFor = (CommunityAdmin) => class extends Component {
    constructor() {
        super();
        this.state = {
            users: [],
            isModalVisible: false,
            newUserEmail: ''
        }
    }

    componentDidMount() {
        let newState = this.state;
        newState.users = UserGateway.getAll();
        this.setState(newState);
    }

    _onShowModal = () => {
        let newState = this.state;
        newState.isModalVisible = true;
        this.setState(newState);
    }

    _onHideModal = () => {
        let newState = this.state;
        newState.isModalVisible = false;
        this.setState(newState);
    }

    _onUpdateEmail = (email) => {
        let newState = this.state;
        newState.newUserEmail = email;
        this.setState(newState);
    }

    _onSubmitEmail = () => {
        this._onHideModal();
        UserGateway.inviteUser(this.state.newUserEmail);
    }

    render() {
        const modal = <CommunityAdminModal
            visible={this.state.isModalVisible}
            onUpdateEmail={this._onUpdateEmail}
            onSubmitEmail={this._onSubmitEmail}
            onHideModal={this._onHideModal}
        />;

        return (
            <CommunityAdmin
                users={this.state.users}
                onShowModal={this._onShowModal}
                modal={modal}
            />
        )
    }
}
