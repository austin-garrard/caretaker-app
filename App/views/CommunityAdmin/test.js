import React, {Component} from 'react';
import renderer from 'react-test-renderer';
import CommunityAdmin from './index.js';
import UserGateway, {Permissions} from '../../data/UserGateway';
describe('Community Admin', () => {

    const users = [
        {name: 'Sarah', identifier: 'sarah@emailprovider.com', permission: Permissions.FOCUS, roles: []},
        {name: 'Caroline', identifier: 'caroline@woahdude.com', permission: Permissions.ADMIN, roles: ['driver', 'coordinator']},
    ]

    beforeEach(() => {
        jest.spyOn(UserGateway, 'getAll')
            .mockReturnValue(users);
    });

    it('snapshot', () => {
        const screen = renderer
            .create(<CommunityAdmin />)

        expect(screen).toMatchSnapshot();
    })

    it('should query the user gateway for all users', () => {
        const screen = renderer
            .create(<CommunityAdmin />)
            .getInstance();

        expect(screen.state.users).toBe(users);
    });

    it('should hide the modal by default', () => {
        const screen = renderer
            .create(<CommunityAdmin />)
            .getInstance();

        expect(screen.state.isModalVisible).toBe(false);
    })

    it('should display the modal when the invite button is pressed', () => {
        const screen = renderer
            .create(<CommunityAdmin />)
            .getInstance();

        screen._onShowModal();

        expect(screen.state.isModalVisible).toBe(true);
    });

    it('should display the modal when the invite button is pressed', () => {
        const screen = renderer
            .create(<CommunityAdmin />)
            .getInstance();

        screen._onHideModal();

        expect(screen.state.isModalVisible).toBe(false);
    });

    it('should hide the modal when the new users email is submitted', () => {
        const screen = renderer
            .create(<CommunityAdmin />)
            .getInstance();

        screen._onSubmitEmail();

        expect(screen.state.isModalVisible).toBe(false);
    });

    it('should update the new user email when typing into the text input', () => {
        const screen = renderer
            .create(<CommunityAdmin />)
            .getInstance();

        screen._onUpdateEmail('email');

        expect(screen.state.newUserEmail).toBe('email');
    })

    it('should invite user through the gateway when the new users email is submitted', () => {
        const spy = jest.spyOn(UserGateway, 'inviteUser')
        const screen = renderer
            .create(<CommunityAdmin />)
            .getInstance();

        screen.state.newUserEmail = 'newUser@gmail.com';
        screen._onSubmitEmail();

        expect(spy).toHaveBeenCalledWith('newUser@gmail.com');
    })
})
