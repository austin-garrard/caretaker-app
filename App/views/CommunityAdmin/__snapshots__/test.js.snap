// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Community Admin snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F5FCFF",
      "justifyContent": "flex-start",
    }
  }
>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={false}
    visible={false}
  >
    <View>
      <TextInput
        onChangeText={[Function]}
        placeholder="Enter email"
      />
      <View
        accessibilityComponentType="button"
        accessibilityLabel={undefined}
        accessibilityTraits={
          Array [
            "button",
          ]
        }
        accessible={true}
        hitSlop={undefined}
        isTVSelectable={true}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#0C42FD",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Submit Email
          </Text>
        </View>
      </View>
      <View
        accessibilityComponentType="button"
        accessibilityLabel={undefined}
        accessibilityTraits={
          Array [
            "button",
          ]
        }
        accessible={true}
        hitSlop={undefined}
        isTVSelectable={true}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#0C42FD",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Cancel
          </Text>
        </View>
      </View>
    </View>
  </Modal>
  <RCTScrollView
    data={
      Array [
        Object {
          "identifier": "sarah@emailprovider.com",
          "name": "Sarah",
          "permission": "focus",
          "roles": Array [],
        },
        Object {
          "identifier": "caroline@woahdude.com",
          "name": "Caroline",
          "permission": "admin",
          "roles": Array [
            "driver",
            "coordinator",
          ],
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onViewableItemsChanged={undefined}
    removeClippedSubviews={true}
    renderItem={[Function]}
    renderScrollComponent={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            Object {
              "padding": 8,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              Sarah
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              sarah@emailprovider.com
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            />
          </View>
          <View>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "textAlign": "center",
                }
              }
            >
              
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            Object {
              "padding": 8,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              Caroline
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              caroline@woahdude.com
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            />
          </View>
          <View>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "textAlign": "center",
                }
              }
            >
              driver, coordinator
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "height": 0,
          }
        }
      />
    </View>
  </RCTScrollView>
  <View
    accessibilityComponentType="button"
    accessibilityLabel={undefined}
    accessibilityTraits={
      Array [
        "button",
      ]
    }
    accessible={true}
    hitSlop={undefined}
    isTVSelectable={true}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#0C42FD",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Invite
      </Text>
    </View>
  </View>
</View>
`;
